File: 0408-Valid Word Abbreviation/0408-Valid Word Abbreviation.cpp
class Solution {
public:
    bool validWordAbbreviation(string word, string abbr) {
        int i = 0, j = 0;
        int num = 0;
        while (i < word.size() && j < abbr.size())
        {
            if (isdigit(abbr[j]))
            {
                num = num * 10 + abbr[j] - '0';
                if (num == 0)
                {
                    return false;
                }                                
            }
            else
            {
                i += num;
                if (i >= word.size() || word[i] != abbr[j])
                {
                    return false;
                }
                
                num = 0;
                ++i;
            }
            ++j;
        }
        
        i += num;
        return i == word.size() && j == abbr.size();
    }
};

File: 0401-0500/0408-Valid Word Abbreviation/0408-Valid Word Abbreviation.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        i = num = 0
        for c in abbr:
            if c.isdigit():
                num = num * 10 + ord(c) - ord('0')
                if num == 0:
                    return False
            else:
                i += num
                if i >= len(word) or word[i] != c:
                    return False
                num = 0
                i += 1
        
        i += num
        return i == len(word)
